### DISCUSSION TITLE
Interested in learning library internals, porting to C++ et al

### DISCUSSION THREAD

**Discussion by tmcopeland** [libsndfile/libsndfile#795] [General]
I've been poking around the source code for the past few weeks. The use of C is certainly impressive. As the title of this post says, I'm interested in developing a port of this library in C++. I know that C wizards might balk at the idea, but I feel that many of C++'s features, particularly those in the newer revisions, would lend themselves nicely to the functionality sought in libsndfile. A few examples:

- Defining generic data structures and functions with templates and overloads to handle different audio encodings (i.e. various integer types, floating-point) and conversions between them
- Using STL containers, streams, buffers, and smart pointers to enable dynamic programming with less headaches
  + Using `std::valarray` and the functions in `<algorithm>` to perform element-wise and/or arithmetic operations on audio data
- Employing C++17's filesystem support for painless I/O across platforms
- Making use of C++20's `std::endian` and C++23's `std::byteswap` to deal with endianness independent of a build system or the preprocessor (at least when these revisions are more widely adopted/supported).

For what it's worth, C11's `_Generic` macro could bring some type-generic functionality to libsndfile.

Obviously, C++ introduces garbage and overhead and produces larger binaries, but I like to think that modern compilers do a good job with optimizations; and after all, most every modern PC has plenty of memory and disk space to go around.

In any case, I'd like to better understand the engineering behind libsndfile. As the authors state themselves, documentation on the library and its code is a bit lacking at the present time. As I've been browsing the code, I've attempted to parse the usage of the synonymous structures `SF_PRIVATE` and `SNDFILE`, as well as the data they contain, and the various functions that operate on them. Is there a chance that any contributors have made a flowchart or something similar to document the library's execution?

My interest in libsndfile came about as a result of taking a Digital Audio Analysis course taught with MATLAB as part of my Audio Technology MA program. MATLAB is far from my favorite platform, but I quite enjoy the ease (modulo compute time) of loading and processing various audio formats as double-precision floats (functionality which, of course, has been replicated in GNU Octave, free of charge).

If anyone has any suggestions or insights about working with digital audio in C/C++ and beyond, I'm all ears!

**Comment by evpobr** [comment]
Hi @tmcopeland . I doubt that it makes sense to rewrite this project in C++, even partially. Too much depends on the fact that it is written in C - supported platforms, build scripts for package managers, and so on. It is very difficult to foresee all the consequences for a project with such a long history.

But if you have any ideas how to do all this beautifully in C++, nothing prevents you from making your own project. If you need some advice on a specific issue, how Libsndfile works, we will be happy to help.

From myself, I advise you to take a look at [libaudiofile](https://github.com/mpruett/audiofile/tree/master/libaudiofile). Maybe you'll find something useful there.
